File: solution/java/characterReplacement.java

import java.util.HashMap;
import java.util.Map;

class Solution {
    public int characterReplacement(String s, int k) {
        Map<Character, Integer> count = new HashMap<>();
        int res = 0;
        for (int maxF = 0, l = 0, h = 0; h < s.length(); h++) {
            count.put(s.charAt(h), count.getOrDefault(s.charAt(h), 0) + 1);
            maxF = Math.max(maxF, count.get(s.charAt(h)));

            while (h - l + 1 - maxF > k) {
                count.put(s.charAt(l), count.getOrDefault(s.charAt(l), 0) - 1);
                l++;
            }

            res = Math.max(res, h - l + 1);

        }
        return res;
    }
}

File: solution/js/characterReplacement.js
var characterReplacement = function (s, k) {
  const count = new Map();
  let res = 0;

  for (let maxF = 0, lo = 0, hi = 0; hi < s.length; hi++) {
    count.set(s[hi], (count.get(s[hi]) || 0) + 1);
    maxF = Math.max(maxF, count.get(s[hi]));

    while (hi - lo + 1 - maxF > k) {
      count.set(s[lo], count.get(s[lo]) - 1);
      lo++;
    }

    res = Math.max(res, hi - lo + 1);
  }

  return res;
};

console.log(characterReplacement("AABABBA", 1));
